server/project_member.go

```diff
@@ -2,18 +2,161 @@ package server
+				}
+				if project.WorkflowType == api.UIWorkflow {
+					return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid workflow type: %s", project.WorkflowType))
+				}
+
+				repoFind := &api.RepositoryFind{ProjectID: &projectID}
+				repo, err := s.RepositoryService.FindRepository(ctx, repoFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Fail to fetch relevant VCS repo, Project ID: %s", c.Param("projectID"))).SetInternal(err)
+				}
+
+				vcsFind := &api.VCSFind{
+					ID: &repo.VCSID,
+				}
+				vcs, err := s.VCSService.FindVCS(ctx, vcsFind)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find VCS for sync project member: %d", repo.VCSID)).SetInternal(err)
+				}
+				if vcs == nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("VCS ID not found: %d", repo.VCSID))
+				}
+
+				gitlabProjectMemberList, err := vcsPlugin.Get("GITLAB_SELF_HOST", vcsPlugin.ProviderConfig{Logger: s.l}).FetchProjectMemberList(ctx,
+					common.OauthContext{
+						ClientID:     vcs.ApplicationID,
+						ClientSecret: vcs.Secret,
+						AccessToken:  repo.AccessToken,
+						RefreshToken: repo.RefreshToken,
+						Refresher:    s.refreshToken(ctx, repo.ID),
+					},
+					vcs.InstanceURL,
+					repo.ExternalID,
+				)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("fail to fetch project member from GitLab, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+				}
+
+				// create or patch project member
+				findProjectMember := &api.ProjectMemberFind{ProjectID: &projectID}
+				bytebaseProjectMemberList, err := s.ProjectMemberService.FindProjectMemberList(ctx, findProjectMember)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("fail to fetch project member in bytebase: %d", projectID)).SetInternal(err)
+				}
+
+				// check whether principal exists in our system. if not exist, create one.
+				for _, projectMember := range gitlabProjectMemberList {
+					findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+					principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+					if err != nil {
+						return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+					}
+
+					if projectMember.State != vcsPlugin.UserStateActive && projectMember.MembershipState != vcsPlugin.UserStateActive {
+						continue
+					}
+
+					if principal == nil {
+						signupInfo := &api.Signup{
+							Name:     projectMember.Name,
+							Email:    projectMember.Email,
+							Password: common.RandomString(20),
+						}
+						createdPrincipal, httpErr := TrySignup(ctx, s, signupInfo, api.PrincipalAuthProviderGitlabSelfHost, c.Get(getPrincipalIDContextKey()).(int))
+						if httpErr != nil {
+							return httpErr
+						}
+						principal = createdPrincipal
+					}
+
+					isProjectMemberExist := false
```

Comment:
nit: if we just created the principal, then it should not exist in the project. And there is no need to go through the for loop